uefi/src/fs/path/path.rs

```diff
@@ -0,0 +1,266 @@
+            Path::new(self)
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use alloc::vec::Vec;
+    use uefi_macros::cstr16;
+
+    #[test]
+    fn from_cstr16() {
+        let source: &CStr16 = cstr16!("\\hello\\foo\\bar");
+        let _path: &Path = source.into();
+        let _path: &Path = Path::new(source);
+    }
+
+    #[test]
+    fn from_cstring16() {
+        let source = CString16::try_from("\\hello\\foo\\bar").unwrap();
+        let _path: &Path = source.as_ref().into();
+        let _path: &Path = Path::new(source.as_ref());
+    }
+
+    #[test]
+    fn components_iter() {
+        let path = Path::new(cstr16!("foo\\bar\\hello"));
+        let components = path.components().collect::<Vec<_>>();
+        let components: Vec<&CStr16> = components.iter().map(|x| x.as_ref()).collect::<Vec<_>>();
+        let expected: &[&CStr16] = &[cstr16!("foo"), cstr16!("bar"), cstr16!("hello")];
+        assert_eq!(components.as_slice(), expected);
+
+        // In case there is a leading slash, it should be ignored.
+        let path = Path::new(cstr16!("\\foo\\bar\\hello"));
+        let components = path.components().collect::<Vec<_>>();
+        let components: Vec<&CStr16> = components.iter().map(|x| x.as_ref()).collect::<Vec<_>>();
+        let expected: &[&CStr16] = &[cstr16!("foo"), cstr16!("bar"), cstr16!("hello")];
+        assert_eq!(components.as_slice(), expected);
+
+        // empty path iteration should be just fine
+        let empty_cstring16 = CString16::try_from("").unwrap();
+        let path = Path::new(empty_cstring16.as_ref());
+        let components = path.components().collect::<Vec<_>>();
+        let expected: &[CString16] = &[];
+        assert_eq!(components.as_slice(), expected);
+
+        // test empty path
+        let _path = Path::new(cstr16!());
+        let path = Path::new(cstr16!(""));
+        let components = path.components().collect::<Vec<_>>();
+        let components: Vec<&CStr16> = components.iter().map(|x| x.as_ref()).collect::<Vec<_>>();
+        let expected: &[&CStr16] = &[];
+        assert_eq!(components.as_slice(), expected);
+
+        // test path that has only root component. Treated as empty path by
+        // the components iterator.
+        let path = Path::new(cstr16!("\\"));
+        let components = path.components().collect::<Vec<_>>();
+        let components: Vec<&CStr16> = components.iter().map(|x| x.as_ref()).collect::<Vec<_>>();
+        let expected: &[&CStr16] = &[];
+        assert_eq!(components.as_slice(), expected);
+    }
+
+    #[test]
+    fn test_parent() {
+        assert_eq!(None, Path::new(cstr16!("")).parent());
+        assert_eq!(None, Path::new(cstr16!("\\")).parent());
+        assert_eq!(
+            Path::new(cstr16!("a\\b")).parent(),
+            Some(PathBuf::from(cstr16!("a"))),
+        );
+        assert_eq!(
+            Path::new(cstr16!("\\a\\b")).parent(),
+            Some(PathBuf::from(cstr16!("a"))),
```

Comment:
Is there a bug here? I would have expected `\\a` instead of just `a`.

Reply:
This is intended and should also be documented on the Components struct. I do specifically handle the root component as a dedicated path component, as all paths are root. We do not have a present working directory here. Because of this simplification, Components doesn't need to be an enum but can be a struct.

Reply:
Ah, yeah, I see. Currently, there is no `.join()` method on a Path/PathBuf.

A Custom PartialEq-impl sounds reasonable